### PR TITLE
Fix validation for tokenized ECv1

### PR DESCRIPTION
**Pull request by brorbw**
Start supporting validation of tokenized ECv1 tokens. Previously they have not been supported.

### CODE REVIEW

**Comment by Lassejoe** [comment]
I have gone through the PR and compared it with https://developers.google.com/pay/api/processors/guides/test-and-validation/sample-tokens#ecv2_4
I have not been able to spot any mistakes.

File: lib/aliquot/validator.rb [hunk 1]
```diff
@@ -166,59 +160,97 @@ class SignedMessageContract < Dry::Validation::Contract
...
+        required(:cryptogram).filled(:str?)
+        required(:eciIndicator).filled(:str?)
+        required(:authMethod).filled(:str?, included_in?: %w[CRYPTOGRAM_3DS])
+      end
+
+      rule(:pan).validate(:integer_string?, :pan?)
+      rule(:eciIndicator).validate(:eci?)
+    end
+
+    # PaymentMethodDetailsSchema = PaymentMethodDetailsContract.new
```
Comment (mt-clearhaus) [inline lib/aliquot/validator.rb:213]:
```suggestion
```

File: lib/aliquot/validator.rb [hunk 2]
```diff
@@ -166,59 +160,97 @@ class SignedMessageContract < Dry::Validation::Contract
...
         required(:paymentMethod).filled(:str?)
-        required(:paymentMethodDetails).filled(:hash).schema(PaymentMethodDetailsContract.schema)
+        required(:paymentMethodDetails).filled(:hash)
         optional(:gatewayMerchantId).filled(:str?)
       end
       rule(:messageExpiration).validate(:integer_string?)
+
+      rule(:paymentMethodDetails).validate do
+        contract = nil
+        if 'ECv1'.eql?(values[:protocolVersion])
```
Comment (mt-clearhaus) [inline lib/aliquot/validator.rb:228]: https://rubystyle.guide/#eql

File: lib/aliquot/validator.rb [hunk 3]
```diff
@@ -166,59 +160,97 @@ class SignedMessageContract < Dry::Validation::Contract
...
+          if 'TOKENIZED_CARD'.eql?(values[:paymentMethod])
+            contract = ECv1_TokenizedPaymentMethodDetailsContract.new
+          else
+            contract =  ECv1_PaymentMethodDetailsContract.new
```
Comment (mt-clearhaus) [inline lib/aliquot/validator.rb:232]:
```suggestion
            contract = ECv1_PaymentMethodDetailsContract.new
```

Comment (mt-clearhaus) [reply lib/aliquot/validator.rb:232]: Perhaps a construction like

```
result = if some_cond
           calc_something
         else
           calc_something_else
         end
```

is preferable.

https://rubystyle.guide/#indent-conditional-assignment

:shrug:

Comment (brorbw) [reply lib/aliquot/validator.rb]: Yeah, it was a bit sloppy written. I prefer to add return for this rather than conditional assignments.

Comment (mt-clearhaus) [reply lib/aliquot/validator.rb]: Me too. I don't know if `return` will work in a Dry rule thing.

File: lib/aliquot/payment.rb [hunk 4]
```diff
@@ -66,6 +66,8 @@ def process
 
       begin
         @message = JSON.parse(decrypt(aes_key, @signed_message[:encryptedMessage]))
+        @message.merge!('threedsCryptogram' => @message.delete('3dsCryptogram')) if @message['3dsCryptogram']
```
Comment (thejspr) [inline lib/aliquot/payment.rb:69]: You could do `@message.transform_keys!('3dsCryptogram' => 'threedsCryptogram')` to rename the key and avoid the conditional

File: lib/aliquot/validator.rb [hunk 5]
```diff
@@ -166,59 +160,95 @@ class SignedMessageContract < Dry::Validation::Contract
...
+          if  values[:authMethod] == 'CRYPTOGRAM_3DS'
+            return ECv2_TokenizedPaymentMethodDetailsContract.new
+          else
+            return ECv2_PaymentMethodDetailsContract.new
+          end
+        end
+      end
       rule(:paymentMethod) do
-        key.failure('must be equal to CARD') unless 'CARD'.eql?(value)
+        if values[:paymentMethodDetails] && values[:paymentMethodDetails].is_a?(Hash)
```
Comment (thejspr) [inline lib/aliquot/validator.rb:242]: `values[:paymentMethodDetails] &&` can be removed as it's covered by `is_a?(Hash)`, but no biggie if you prefer to be explicit.

File: lib/aliquot/validator.rb [hunk 6]
```diff
@@ -166,59 +160,95 @@ class SignedMessageContract < Dry::Validation::Contract
...
+          if '3DS'.eql?(values[:paymentMethodDetails] && values[:paymentMethodDetails]['authMethod']) # Tokenized ECv1
```
Comment (thejspr) [inline lib/aliquot/validator.rb:243]: `values[:paymentMethodDetails] &&` is covered in the parent `if` statement

File: spec/lib/aliquot/validator_spec.rb [hunk 7]
```diff
@@ -78,426 +77,523 @@
...
-      let(:input)  { token }
 
-      it 'must exist' do
-        token.delete('intermediateSigningKey')
-        is_expected.to dissatisfy_schema(schema, {intermediateSigningKey: ['is missing']})
+    context 'tokenized' do
+      let(:schema) { Aliquot::Validator::ECv2_TokenizedPaymentMethodDetailsContract.new }
+      let(:generator) { AliquotPay.new(protocol_version: :ECv2, type: :browser) }
+      let(:token)     { generator.token }
+      # include_examples 'Validator Spec'
```
Comment (thejspr) [inline spec/lib/aliquot/validator_spec.rb:580]: Should this shared example be enabled?

**Comment by thejspr** [comment]
A few small suggestions from me, otherwise it looks solid.

File: lib/aliquot/payment.rb (hunk 4 above)
Comment (brorbw) [reply lib/aliquot/payment.rb]: I have to reverse this as it appears that it is not supported in ruby 2.7

Comment (thejspr) [reply lib/aliquot/payment.rb]: My bad, it was introduced in 2.5, but looks like it only worked with a block then.

**Review by thejspr — APPROVED** [review]

File: Gemfile [hunk 8]
```diff
@@ -1,3 +1,5 @@
 source 'https://rubygems.org'
 
+gem 'aliquot-pay', :git => 'https://github.com/clearhaus/aliquot-pay', :branch => 'make-tokens-consistent-with-googles-sample-tokens'
```
[the remark this answers is not in the document]
Comment (brorbw) [reply Gemfile:3]: Yes indeed it should but with the mutual dependency the automated tests will fail without.

[the remark this answers is not in the document]
Comment (cbobach) [reply Gemfile:3]: It was just to highlight that this was still hanging around even though you got a :heavy_check_mark:

**Review by thejspr — APPROVED** [review]

File: lib/aliquot/payment.rb [hunk 9]
```diff
@@ -66,6 +66,9 @@ def process
 
       begin
         @message = JSON.parse(decrypt(aes_key, @signed_message[:encryptedMessage]))
+        @message['paymentMethodDetails'].merge!(
+          'threedsCryptogram' => @message['paymentMethodDetails']
+          .delete('3dsCryptogram')) if @message['paymentMethodDetails']['3dsCryptogram']
```
Comment (mt-clearhaus) [inline lib/aliquot/payment.rb:71]: This is IMO not easily readable.

Comment (mt-clearhaus) [reply lib/aliquot/payment.rb:71]:
```suggestion

        details = @message['paymentMethodDetails']
        if details['3dsCryptogram']
          details['threedsCryptogram'] = details.delete('3dsCryptogram')
        end
```

File: lib/aliquot/validator.rb [hunk 10]
```diff
@@ -166,59 +160,96 @@ class SignedMessageContract < Dry::Validation::Contract
...
+             ECv2_PaymentMethodDetailsContract.new
+          end
+        end
+        contract.call(values[:paymentMethodDetails])
+      end
+
       rule(:paymentMethod) do
-        key.failure('must be equal to CARD') unless 'CARD'.eql?(value)
+        if values[:paymentMethodDetails].is_a?(Hash)
+          if '3DS'.eql?(values[:paymentMethodDetails]['authMethod']) # Tokenized ECv1
```
Comment (mt-clearhaus) [inline lib/aliquot/validator.rb:244]: https://rubystyle.guide/#eql

Comment (mt-clearhaus) [reply lib/aliquot/validator.rb:244]:
```suggestion
          if values[:paymentMethodDetails]['authMethod'] == '3DS' # Tokenized ECv1
```

File: lib/aliquot/validator.rb [hunk 11]
```diff
@@ -283,7 +314,32 @@ class Error < ::Aliquot::Error; end
...
+          else
+            Aliquot::Validator::ECv1_PaymentMethodDetailsContract.new
+          end
+        else
+          if tokenized
+            Aliquot::Validator::ECv2_TokenizedPaymentMethodDetailsContract.new
+          else
+            Aliquot::Validator::ECv2_PaymentMethodDetailsContract.new
+          end
+        end
```
Comment (mt-clearhaus) [inline lib/aliquot/validator.rb:342]:
```suggestion
          if version == 'ECv1'
            if tokenized
              Aliquot::Validator::ECv1_TokenizedPaymentMethodDetailsContract.new
            else
              Aliquot::Validator::ECv1_PaymentMethodDetailsContract.new
            end
          else
            if tokenized
              Aliquot::Validator::ECv2_TokenizedPaymentMethodDetailsContract.new
            else
              Aliquot::Validator::ECv2_PaymentMethodDetailsContract.new
            end
          end
```

**Review by mt-clearhaus — COMMENTED** [review]
Just a few things I stumbled upon. Only syntax stuff.

File: lib/aliquot/payment.rb (hunk 9 above)
Comment (mt-clearhaus) [reply lib/aliquot/payment.rb:71]:
```suggestion

        d = @message['paymentMethodDetails']
        d['threedsCryptogram'] = d.delete('3dsCryptogram') if details['3dsCryptogram']
```

**Pull request merged**